File: bingo/withdraw.py
import os
from decimal import Decimal
from telegram import Update, ReplyKeyboardRemove, ReplyKeyboardMarkup, KeyboardButton
from telegram.ext import ContextTypes
from telegram.constants import ParseMode
from asgiref.sync import sync_to_async

from .models import Player


MIN_WITHDRAW = Decimal("100.00")
WITHDRAW_METHODS = [
    ("telebirr", "Telebirr"),
    ("cbe_birr", "CBE Birr"),
    ("boa", "BOA"),
    ("cbe", "CBE"),
]


async def start_withdraw(update: Update, context: ContextTypes.DEFAULT_TYPE) -> None:
    # Initialize withdraw conversation state
    context.user_data["withdraw_state"] = "await_amount"
    context.user_data["withdraw"] = {}
    await update.effective_message.reply_text(
        "Please enter the amount you wish to withdraw.", reply_markup=ReplyKeyboardRemove()
    )


async def handle_withdraw_amount(update: Update, context: ContextTypes.DEFAULT_TYPE) -> None:
    # Backward compatibility with previous flag
    if context.user_data.get("await_withdraw_amount") and not context.user_data.get("withdraw_state"):
        context.user_data["withdraw_state"] = "await_amount"
        context.user_data["withdraw"] = {}

    state = context.user_data.get("withdraw_state")
    if not state:
        return

    text = (update.message.text or "").strip()
    # Simple cancel hook
    if text.lower() in {"cancel", "/cancel"}:
        context.user_data.pop("withdraw_state", None)
        context.user_data.pop("withdraw", None)
        context.user_data.pop("await_withdraw_amount", None)
        await update.message.reply_text("Withdraw cancelled.")
        return

    # Step 1: amount
    if state == "await_amount":
        try:
            amt = Decimal(text)
        except Exception:
            await update.message.reply_text("Please enter a valid number amount.")
            return

        if amt < MIN_WITHDRAW:
            await update.message.reply_text(
                f"Withdraw amount must be greater than or equal to {MIN_WITHDRAW:.0f}"
            )
            return

        user = update.effective_user
        if not user:
            await update.message.reply_text("Unable to process your request right now.")
            return

        def load_player():
            p = Player.objects.filter(telegram_id=user.id).first()
            if not p:
                return None, None, None
            bal = p.wallet or Decimal("0")
            phone = p.phone or "-"
            return p, bal, phone

        p, before, phone = await sync_to_async(load_player)()
        if before is None:
            await update.message.reply_text("Please register first using /start and share your phone number.")
            return

        if before < amt:
            await update.message.reply_text(
                f"Insufficient fund. user: {user.id}, amount: {amt:.1f}."
            )
            return

        # Store for next steps
        context.user_data["withdraw"] = {"amount": amt, "balance": before, "phone": phone}
        context.user_data["withdraw_state"] = "await_method"

        # Ask for method
        kb = ReplyKeyboardMarkup(
            [[KeyboardButton(lbl) for _key, lbl in WITHDRAW_METHODS], [KeyboardButton("Cancel")]],
            resize_keyboard=True,
            one_time_keyboard=True,
            input_field_placeholder="Choose withdraw method",
        )
        await update.message.reply_text("Please choose your withdraw method:", reply_markup=kb)
        return

    # Step 2: method
    if state == "await_method":
        norm = text.strip().lower()
        matched = None
        for key, lbl in WITHDRAW_METHODS:
            if norm in {key, lbl.lower()}:
                matched = key
                display = lbl
                break
        if not matched:
            await update.message.reply_text("Please choose a valid method: Telebirr, CBE Birr, BOA, or CBE.")
            return

        context.user_data["withdraw"]["method"] = matched
        context.user_data["withdraw"]["method_label"] = display
        context.user_data["withdraw_state"] = "await_account"
        await update.message.reply_text(
            f"Enter your {display} account/phone to receive the withdrawal:",
            reply_markup=ReplyKeyboardRemove(),
        )
        return

    # Step 3: account details
    if state == "await_account":
        account = text
        info = context.user_data.get("withdraw") or {}
        amt = info.get("amount")
        before = info.get("balance")
        phone = info.get("phone") or "-"
        method_label = info.get("method_label") or "-"

        # Clear state
        context.user_data.pop("withdraw_state", None)
        context.user_data.pop("withdraw", None)
        context.user_data.pop("await_withdraw_amount", None)
        user = update.effective_user
        if not user or amt is None or before is None:
            await update.message.reply_text("Unable to process your request right now.")
            return

        # Notify entertainer for manual processing
        ent_chat = os.getenv("ENTERTAINER_ID")
        ent_id_val = None
        if ent_chat:
            try:
                ent_id_val = int(ent_chat)
            except Exception:
                ent_id_val = None

        username = (update.effective_user.username if update.effective_user else None) or "-"
        if ent_id_val is not None:
            try:
                await context.bot.send_message(
                    chat_id=ent_id_val,
                    text=(
                        "Withdrawal Request\n"
                        f"User: @{username} (id: {user.id})\n"
                        f"Phone: {phone}\n"
                        f"Amount: {amt:.2f} ETB\n"
                        f"Method: {method_label}\n"
                        f"Account: {account}\n"
                        f"Current Balance: {before:.2f} ETB\n"
                    ),
                    parse_mode=ParseMode.HTML,
                )
            except Exception:
                pass

        await update.message.reply_text(
            (
                "Your withdrawal request has been received and is being processed by admin.\n"
                f"Requested Amount: {amt:.2f} ETB\n"
                f"Method: {method_label}\n"
                f"Account: {account}\n"
                f"Current Balance: {before:.2f} ETB"
            ),
            parse_mode=ParseMode.HTML,
        )
        return

File: server/bot/withdraw.js
// Withdraw handler — port of bingo/withdraw.py
const { PrismaClient } = require("@prisma/client");
const { Decimal } = require("decimal.js");
const { notifyEntertainers, getEntertainerIds } = require("./entertainer");
const prisma = new PrismaClient();

const MIN_WITHDRAW = new Decimal(100);
const METHODS = [
  { key: "telebirr", label: "Telebirr / ቴሌብር" },
  { key: "cbe_birr", label: "CBE Birr / ሲቢኢ ብር" },
  { key: "boa", label: "BOA / ቢኦኤ" },
  { key: "cbe", label: "CBE / ሲቢኢ" },
];

function setupWithdraw(bot, userState) {
  // Helper to reset conversation state when starting withdraw
  function resetState(state) {
    // Clear deposit flow
    state.lastDepositMethod = null;
    state.depositAmount = null;
    state.awaitingDepositAmount = false;
    state.awaitingDepositReceipt = false;
    // Clear transfer flow
    state.transferStep = null;
    state.transfer = null;
    // Clear convert flow
    state.convertStep = null;
    state.convert = null;
    // Clear report flow
    state.reportStep = null;
    state.report = null;
  }

  bot.onText(/\/withdraw/, async (msg) => {
    const chatId = msg.chat.id;
    const tid = msg.from.id;

    if (!userState.has(tid)) userState.set(tid, {});
    const state = userState.get(tid);
    resetState(state); // Cancel any previous unfinished command
    state.withdrawStep = "amount";
    state.withdraw = {};

    await bot.sendMessage(
      chatId,
      "Please enter the amount you wish to withdraw in ETB.\n" +
        "እባክዎ ማውጣት የሚፈልጉትን የገንዘብ መጠን በብር ያስገቡ።\n\n" +
        `Minimum withdraw amount is ${MIN_WITHDRAW.toFixed(0)} ETB.\n` +
        `የማውጣት ከፍተኛ መጠን ${MIN_WITHDRAW.toFixed(0)} ብር ነው።\n\n` +
        "Type 'Cancel' to stop. / ለመመለስ 'Cancel' ብለው ይፃፉ።",
      {
        reply_markup: { remove_keyboard: true },
      },
    );
  });

  bot.on("message", async (msg) => {
    if (!msg.text || msg.text.startsWith("/")) return;
    const chatId = msg.chat.id;
    const tid = msg.from.id;
    const state = userState.get(tid);
    if (!state || !state.withdrawStep) return;

    const text = msg.text.trim();
    if (text.toLowerCase() === "cancel" || text === "Cancel / ሰርዝ" || text === "ሰርዝ") {
      userState.delete(tid);
      await bot.sendMessage(chatId, "Withdraw cancelled.");
      return;
    }

    if (state.withdrawStep === "amount") {
      let amount;
      try {
        amount = new Decimal(text);
      } catch (_) {
        amount = null;
      }
      if (!amount) {
        await bot.sendMessage(chatId, "Please enter a valid number amount.");
        return;
      }
      if (amount.lt(MIN_WITHDRAW)) {
        await bot.sendMessage(
          chatId,
          `Withdraw amount must be greater than or equal to ${MIN_WITHDRAW.toFixed(0)}`,
        );
        return;
      }

      const player = await prisma.player.findUnique({
        where: { telegramId: BigInt(tid) },
      });
      if (!player) {
        await bot.sendMessage(
          chatId,
          "Please register first using /start and share your phone number.",
        );
        userState.delete(tid);
        return;
      }

      const mainBalance = new Decimal(player.wallet.toString());
      const playBalance = new Decimal(player.gift.toString());
      if (mainBalance.lt(amount)) {
        let message = `❌ Withdrawal Failed — ማውጣት አልተቻለም ❌\n\n`;
        message += `Your main wallet balance: ${mainBalance.toFixed(2)} ETB\n`;
        message += `የዋና ቦርሳዎ ቀሪ ሂሳብ: ${mainBalance.toFixed(2)} ብር\n\n`;
        message += `Requested amount: ${amount.toFixed(2)} ETB\n`;
        message += `የጠየቁት መጠን: ${amount.toFixed(2)} ብር\n\n`;
        message += `⚠️ Your main wallet is less than you asked.\n`;
        message += `⚠️ የዋና ቦርሳዎ ከጠየቁት ያነሰ ነው።\n\n`;
        if (playBalance.gt(0)) {
          message += `ℹ️ Play wallet balance: ${playBalance.toFixed(2)} ETB\n`;
          message += `ℹ️ የጨዋታ ቦርሳ ቀሪ ሂሳብ: ${playBalance.toFixed(2)} ብር\n\n`;
          message += `Note: Play wallet can only be used for playing, not for withdrawal.\n`;
          message += `ማሳሰቢያ: የጨዋታ ቦርሳ ለመጫወት ብቻ ነው የሚሆነው፣ ለማውጣት አይቻልም።`;
        }
        await bot.sendMessage(chatId, message);
        userState.delete(tid);
        return;
      }

      state.withdraw = { amount, balance: mainBalance, phone: player.phone || "-" };
      state.withdrawStep = "method";

      await bot.sendMessage(chatId, "Please choose your withdraw method:", {
        reply_markup: {
          keyboard: [
            METHODS.map((m) => ({ text: m.label })),
            [{ text: "Cancel / ሰርዝ" }],
          ],
          resize_keyboard: true,
          one_time_keyboard: true,
          input_field_placeholder: "Choose withdraw method",
        },
      });
      return;
    }

    if (state.withdrawStep === "method") {
      const norm = text.toLowerCase();
      const matched = METHODS.find(
        (m) => m.key === norm || m.label.toLowerCase() === norm || m.label === text || m.label.split(" / ")[0].toLowerCase() === norm,
      );
      if (!matched) {
        await bot.sendMessage(
          chatId,
          "Please choose a valid method: Telebirr, CBE Birr, BOA, or CBE.",
        );
        return;
      }
      state.withdraw.method = matched.key;
      state.withdraw.methodLabel = matched.label;
      state.withdrawStep = "account";
      await bot.sendMessage(
        chatId,
        `Enter your ${matched.label} account/phone to receive the withdrawal:`,
        {
          reply_markup: { remove_keyboard: true },
        },
      );
      return;
    }

    if (state.withdrawStep === "account") {
      const account = text;
      const info = state.withdraw || {};
      const amount = info.amount;
      const balance = info.balance;
      const phone = info.phone || "-";
      const methodLabel = info.methodLabel || "-";
      const method = info.method || "";

      userState.delete(tid);

      if (!amount || !balance) {
        await bot.sendMessage(
          chatId,
          "Unable to process your request right now.",
        );
        return;
      }

      const username = msg.from.username || "-";

      let withdrawRequestId = null;
      try {
        const player = await prisma.player.findUnique({
          where: { telegramId: BigInt(tid) },
        });
        if (player) {
          const walletDec = new Decimal(player.wallet.toString());
          if (walletDec.lt(amount)) {
            await bot.sendMessage(chatId, "❌ Insufficient balance.");
            return;
          }

          // Hold pattern: deduct immediately, refund on reject
          const result = await prisma.$transaction(async (tx) => {
            await tx.player.update({
              where: { id: player.id },
              data: { wallet: { decrement: amount.toNumber() } },
            });
            await tx.transaction.create({
              data: {
                playerId: player.id,
                kind: "withdraw_hold",
                amount: amount.negated().toNumber(),
                note: "Withdraw request hold (pending admin approval)",
              },
            });
            return tx.withdrawRequest.create({
              data: {
                playerId: player.id,
                telegramId: BigInt(tid),
                amount: amount.toNumber(),
                method,
                account,
                status: "pending",
              },
            });
          });
          withdrawRequestId = result.id;
        }
      } catch (err) {
        console.error("withdraw request persist error:", err);
      }

      const entertainerIds = getEntertainerIds();
      if (entertainerIds.length > 0 && withdrawRequestId) {
        const message =
          `🏧 Withdrawal Request #${withdrawRequestId}\n` +
          `━━━━━━━━━━━━━━━━━━\n` +
          `User: @${username} (id: ${tid})\n` +
          `Phone: ${phone}\n` +
          `Amount: ${amount.toFixed(2)} ETB\n` +
          `Method: ${methodLabel}\n` +
          `Account: ${account}\n` +
          `Current Balance: ${balance.toFixed(2)} ETB\n`;

        await notifyEntertainers(bot, message, {
          reply_markup: {
            inline_keyboard: [
              [
                {
                  text: "✅ Approve",
                  callback_data: `approve_withdraw:${withdrawRequestId}`,
                },
                {
                  text: "❌ Reject",
                  callback_data: `reject_withdraw:${withdrawRequestId}`,
                },
              ],
            ],
          },
        });
      }

      await bot.sendMessage(
        chatId,
        "Your withdrawal request has been received and is being processed by admin.\n" +
          `Requested Amount: ${amount.toFixed(2)} ETB\n` +
          `Method: ${methodLabel}\n` +
          `Account: ${account}\n` +
          `Current Balance: ${balance.toFixed(2)} ETB`,
      );
    }
  });
}

module.exports = { setupWithdraw };
